cpp/cycles_module/algorithm/cycles.cpp

```diff
@@ -0,0 +1,201 @@
+
+std::uint64_t NodeState::GetDepth(std::uint64_t node_id) const { return depth[node_id]; }
+
+void FindNonSpanningTreeEdges(std::uint64_t node_id, const mg_graph::GraphView<> &graph, NodeState *state,
+                              std::set<std::pair<uint64_t, uint64_t>> *non_st_edges) {
+  std::unordered_set<std::uint64_t> unique_neighbour;
+  state->SetVisited(node_id);
+  for (const auto &neigh : graph.Neighbours(node_id)) {
+    auto next_id = neigh.node_id;
+
+    // Check if is returning edge or already visited neighbour
+    if (next_id == state->GetParent(node_id) || unique_neighbour.find(next_id) != unique_neighbour.end()) {
+      continue;
+    }
+
+    unique_neighbour.insert(next_id);
+
+    if (state->IsVisited(next_id)) {
+      auto sorted_edge = std::minmax(node_id, next_id);
+      non_st_edges->insert(sorted_edge);
+      continue;
+    }
+
+    // Set depth and parent for the next ST iteration
+    state->SetParent(next_id, node_id);
+    state->SetDepth(next_id, state->GetDepth(node_id) + 1);
+    FindNonSpanningTreeEdges(next_id, graph, state, non_st_edges);
+  }
+}
+
+void FindFundamentalCycles(const std::set<std::pair<std::uint64_t, std::uint64_t>> &non_st_edges,
+                           const NodeState &state, std::vector<std::vector<std::uint64_t>> *fundamental_cycles) {
+  for (const auto [from, to] : non_st_edges) {
+    fundamental_cycles->emplace_back(FindFundametalCycle(from, to, state));
+  }
+}
+
+std::vector<std::uint64_t> FindFundametalCycle(std::uint64_t node_a, std::uint64_t node_b, const NodeState &state) {
+  std::vector<std::uint64_t> cycle;
+  if (state.depth[node_a] < state.depth[node_b]) {
+    std::swap(node_a, node_b);
+  }
+
+  // climb until a and b reach the same depth:
+  //
+  //       ()               ()
+  //      /  \             /  \.
+  //    ()   (b)   -->   (a)  (b)
+  //   /  \             /   \.
+  // (a)  ()           ()   ()
+  cycle.emplace_back(node_a);
+  while (state.depth[node_a] > state.depth[node_b]) {
+    node_a = state.parent[node_a];
+    cycle.emplace_back(node_a);
+  }
+
+  if (node_a != node_b) {
+    throw std::runtime_error("There should be no cross edges in DFS tree of an undirected graph.");
+  }
+
+  // Close the cycle
+  cycle.emplace_back(cycle[0]);
+  return cycle;
+}
+
+void CombineCycles(int mask, const std::vector<std::vector<std::uint64_t>> &fundamental_cycles,
+                   const mg_graph::GraphView<> &graph, std::vector<std::vector<mg_graph::Node<>>> *cycles) {
+  std::map<std::pair<std::uint64_t, std::uint64_t>, std::uint64_t> edge_cnt;
+  for (std::size_t i = 0; i < fundamental_cycles.size(); ++i) {
+    if ((mask & (1 << i)) == 0) continue;
+    for (std::size_t j = 1; j < fundamental_cycles[i].size(); ++j) {
+      auto edge = std::minmax(fundamental_cycles[i][j], fundamental_cycles[i][j - 1]);
+      edge_cnt[edge]++;
+    }
+  }
+
+  std::map<std::uint64_t, std::vector<std::uint64_t>> adj_list;
+  std::set<std::uint64_t> nodes;
```

Comment:
Use `unordered_map` and `unordered_set` if you don't care about the order, and it seems you don't.